src/parser/range.rs

```diff
@@ -420,16 +420,32 @@ where
                         // because we've already done it on look_ahead_input.
                         unreachable!();
                     } else {
-                        input.reset(look_ahead_input);
-                        *to_consume += 1;
+                        // Clone because we need look_ahead_input below to calculate how many bytes were consumed
+                        input.reset(look_ahead_input.clone());
+
+                        // Advance the stream
                         if input.uncons().is_err() {
                             unreachable!();
                         }
+
+                        // Calculate how many bytes input.uncons() advanced the stream
+                        *to_consume += input.distance(&look_ahead_input);
                     }
                 }
-                Err(e) => {
-                    input.reset(before);
-                    return wrap_stream_error(input, e);
+                Err(first_error) => {
+                    // Clone because we need look_ahead_input below to calculate how many bytes were consumed
+                    input.reset(look_ahead_input.clone());
+                    
+                    // See if we can advance anyway
+                    if let Err(_) = input.uncons() {
+                        // Return the original error if uncons failed
```

Comment:
The error branch needs to reset the input to `before`, otherwise partial parsing will fail (I need to add a test for this specifically though).

Reply:
I added the test in another branch 6c181905c53edf68d8ed9e12cfa87280f179a411, but it's failing in a weird non-deterministic way. Hopefully I'll have some more time this evening to work on it.

Reply:
That's the point of the test :) It generates partial reads to test that the parser can cope. It does print out which reads that caused it to fail though. So you can copy those from stderr and define `let seq = vec![PartialOp(x), PartialOp(y), ..];` inside the test and then run it again to see the same failure. Then it is possible to start up a debugger and check what is wrong.